File: packages/orchestrator/src/memory/store.ts
import { appendFileSync, mkdirSync, readFileSync, writeFileSync, existsSync } from "fs";
import { join, dirname } from "path";
import { fileURLToPath } from "url";
import { readEvents } from "../events/store.js";
import { extractEntities } from "./extract.js";
import type { RequestMemory, ClarificationQA } from "./types.js";
import type { LLMClient } from "../llm/doubao.js";

const __filename = fileURLToPath(import.meta.url);
// packages/orchestrator/src/memory → ../../../../memory
const DEFAULT_MEMORY_DIR = join(dirname(__filename), "../../../../memory");
const MEMORY_FILE = "store.jsonl";

function getMemoryFile(): string {
  const dir = process.env.MEMORY_DIR ?? DEFAULT_MEMORY_DIR;
  if (!existsSync(dir)) mkdirSync(dir, { recursive: true });
  return join(dir, MEMORY_FILE);
}

// ────────────────────────────────────────────────────────────
// 读 / 写
// ────────────────────────────────────────────────────────────

export function appendMemory(mem: RequestMemory): void {
  appendFileSync(getMemoryFile(), JSON.stringify(mem) + "\n", "utf8");
}

export function readAllMemories(): RequestMemory[] {
  const file = getMemoryFile();
  if (!existsSync(file)) return [];
  return readFileSync(file, "utf8")
    .split("\n").filter(Boolean)
    .map((line) => {
      try { return JSON.parse(line) as RequestMemory; } catch { return null; }
    })
    .filter((m): m is RequestMemory => m !== null);
}

/** 用于回填覆盖（如重新跑抽取）：按 runId 去重，新条目替换旧条目 */
export function upsertMemory(mem: RequestMemory): void {
  const all = readAllMemories().filter((m) => m.runId !== mem.runId);
  all.push(mem);
  writeFileSync(getMemoryFile(), all.map((m) => JSON.stringify(m)).join("\n") + "\n", "utf8");
}

// ────────────────────────────────────────────────────────────
// 从 events/<runId>.jsonl 重建一个 RequestMemory（除 entities 外）
// ────────────────────────────────────────────────────────────

interface EventLike {
  type: string;
  ts: number;
  payload: Record<string, unknown>;
}

export interface PartialMemory {
  runId: string;
  ts: number;
  summary: string;
  skillUsed: string;
  changedFiles: string[];
  clarifications: ClarificationQA[];
  outcome: "verified" | "failed";
}

/** 从某个 runId 的 jsonl 提取除 entities 之外的所有字段。
 *  无法重建（缺关键事件）时返回 null。 */
export function buildPartialFromEvents(runId: string): PartialMemory | null {
  const events = readEvents(runId) as unknown as EventLike[];
  if (events.length === 0) return null;

  const start = events[0];

  // clarify.done 拿 summary + answers
  const clarifyDone = events.find((e) => e.type === "clarify.done");
  const req = clarifyDone?.payload?.req as Record<string, unknown> | undefined;
  const summary = String(req?.summary ?? "(no summary)");
  const answers = (req?.answers as Record<string, string>) ?? {};

  // 从 clarify.questions 事件取 q→aspect 映射（兼容旧 run：无该事件或无 aspect 字段时落回 "other"）
  const clarifyQuestions = events.find((e) => e.type === "clarify.questions");
  const rawQs = (clarifyQuestions?.payload?.questions ?? []) as Array<{ q?: string; aspect?: string } | string>;
  const aspectMap = new Map<string, string>();
  for (const q of rawQs) {
    if (typeof q === "string") aspectMap.set(q, "other");
    else if (q?.q) aspectMap.set(q.q, q.aspect ?? "other");
  }
  const clarifications: ClarificationQA[] = Object.entries(answers).map(([q, a]) => ({
    q, a, aspect: aspectMap.get(q) ?? "other",
  }));

  // plan.done 拿 skillName
  const planDone = events.find((e) => e.type === "plan.done");
  const skillUsed = String(planDone?.payload?.skillName ?? "");

  // commit.done 优先（含 git 实际 commit 的文件），降级到 code.done.files / locate.done.files
  const commitDone = events.find((e) => e.type === "commit.done");
  const codeDone = events.find((e) => e.type === "code.done");
  const locateDone = events.find((e) => e.type === "locate.done");
  let changedFiles: string[] = [];
  if (commitDone?.payload?.patchCount && codeDone) {
    changedFiles = (codeDone.payload.files as string[]) ?? [];
  } else if (codeDone) {
    changedFiles = (codeDone.payload.files as string[]) ?? [];
  } else if (locateDone) {
    const files = locateDone.payload.files as Array<{ path: string }> | undefined;
    changedFiles = files?.map((f) => f.path) ?? [];
  }

  // outcome：看是否有 run.completed 和 verify.done
  const completed = events.some((e) => e.type === "run.completed");
  const verified = events.some((e) => e.type === "verify.done");
  const outcome: "verified" | "failed" = (completed && verified) ? "verified" : "failed";

  if (!skillUsed || !summary) return null;

  return { runId, ts: start.ts, summary, skillUsed, changedFiles, clarifications, outcome };
}

// ────────────────────────────────────────────────────────────
// 持久化入口：run 完成时由 orchestrator 调用一次
// ────────────────────────────────────────────────────────────

export async function persistMemory(runId: string, llm: LLMClient): Promise<RequestMemory | null> {
  const partial = buildPartialFromEvents(runId);
  if (!partial) {
    console.warn(`[memory] 无法从 events 重建 partial memory for runId=${runId}`);
    return null;
  }

  const entities = await extractEntities(llm, {
    runId,
    summary: partial.summary,
    skillUsed: partial.skillUsed,
    changedFiles: partial.changedFiles,
    clarifications: partial.clarifications,
  });

  const memory: RequestMemory = { ...partial, entities };
  upsertMemory(memory);
  return memory;
}

File: packages/orchestrator/orchestrator/memory/store.py
from __future__ import annotations

import json
import os
from pathlib import Path

from orchestrator.events.store import read_events
from orchestrator.llm.doubao import LLMClient
from orchestrator.memory.extract import extract_entities
from orchestrator.types import ClarificationQA, ExtractedEntities, RequestMemory

ROOT = Path(__file__).resolve().parents[4]
MEMORY_DIR = Path(os.getenv("MEMORY_DIR", str(ROOT / "memory")))
MEMORY_FILE = "store.jsonl"


def _file() -> Path:
    MEMORY_DIR.mkdir(parents=True, exist_ok=True)
    return MEMORY_DIR / MEMORY_FILE


def _entities_from_dict(data: dict) -> ExtractedEntities:
    return ExtractedEntities(
        domainObjects=data.get("domainObjects", []),
        operations=data.get("operations", []),
        affectedLayers=data.get("affectedLayers", []),
        fieldsAdded=data.get("fieldsAdded", []),
        uiSurfaces=data.get("uiSurfaces", []),
    )


def _memory_from_dict(data: dict) -> RequestMemory:
    return RequestMemory(
        runId=data["runId"],
        ts=data["ts"],
        summary=data["summary"],
        skillUsed=data.get("skillUsed"),
        mode=data.get("mode", "skill"),
        genericPlanFiles=data.get("genericPlanFiles"),
        entities=_entities_from_dict(data.get("entities", {})),
        changedFiles=data.get("changedFiles", []),
        clarifications=[ClarificationQA(**c) for c in data.get("clarifications", [])],
        outcome=data.get("outcome", "failed"),
        repoNwo=data.get("repoNwo"),
    )


def read_all_memories() -> list[RequestMemory]:
    path = _file()
    if not path.exists():
        return []
    out: list[RequestMemory] = []
    for line in path.read_text(encoding="utf-8").splitlines():
        try:
            out.append(_memory_from_dict(json.loads(line)))
        except Exception:
            pass
    return out


def upsert_memory(mem: RequestMemory) -> None:
    all_mem = [m for m in read_all_memories() if m.runId != mem.runId]
    all_mem.append(mem)
    lines = []
    for m in all_mem:
        data = {
            **m.__dict__,
            "entities": m.entities.__dict__,
            "clarifications": [c.__dict__ for c in m.clarifications],
        }
        lines.append(json.dumps(data, ensure_ascii=False))
    _file().write_text("\n".join(lines) + "\n", encoding="utf-8")


def build_partial_from_events(run_id: str) -> dict | None:
    events = read_events(run_id)
    if not events:
        return None
    clarify_done = next((e for e in events if e.type == "clarify.done"), None)
    req = (clarify_done.payload.get("req") if clarify_done else {}) or {}
    summary = str(req.get("summary", "(no summary)"))
    answers = req.get("answers") or {}
    clarify_questions = next((e for e in events if e.type == "clarify.questions"), None)
    raw_qs = (clarify_questions.payload.get("questions") if clarify_questions else []) or []
    aspect_map: dict[str, str] = {}
    for q in raw_qs:
        if isinstance(q, str):
            aspect_map[q] = "other"
        elif isinstance(q, dict) and q.get("q"):
            aspect_map[str(q["q"])] = str(q.get("aspect", "other"))
    clarifications = [ClarificationQA(q=q, a=a, aspect=aspect_map.get(q, "other")) for q, a in answers.items()]
    plan_done = next((e for e in events if e.type == "plan.done"), None)
    plan_generic = next((e for e in events if e.type == "plan.generic"), None)
    mode = "generic" if plan_generic else "skill"
    skill_used = None if plan_generic else (str(plan_done.payload.get("skillName", "")) if plan_done else "")
    if mode == "skill" and not skill_used:
        return None
    generic_plan_files = [str(f.get("path", "")) for f in (plan_generic.payload.get("files", []) if plan_generic else []) if f.get("path")]
    code_done = next((e for e in events if e.type == "code.done"), None)
    locate_done = next((e for e in events if e.type == "locate.done"), None)
    changed_files = code_done.payload.get("files", []) if code_done else []
    if not changed_files and locate_done:
        changed_files = [f.get("path") for f in locate_done.payload.get("files", []) if f.get("path")]
    completed = any(e.type == "run.completed" for e in events)
    verified = any(e.type == "verify.done" for e in events)
    return {
        "runId": run_id,
        "ts": events[0].ts,
        "summary": summary,
        "mode": mode,
        "skillUsed": skill_used,
        "genericPlanFiles": generic_plan_files or None,
        "changedFiles": changed_files,
        "clarifications": clarifications,
        "outcome": "verified" if completed and verified else "failed",
    }


async def persist_memory(run_id: str, llm: LLMClient, repo_nwo: str | None = None) -> RequestMemory | None:
    partial = build_partial_from_events(run_id)
    if not partial:
        return None
    entities = await extract_entities(
        llm,
        run_id=run_id,
        summary=partial["summary"],
        skill_used=partial["skillUsed"],
        changed_files=partial["changedFiles"],
        clarifications=partial["clarifications"],
    )
    mem = RequestMemory(**partial, entities=entities, repoNwo=repo_nwo)
    upsert_memory(mem)
    return mem

